25 social login

leehyunk6310 opened the pull request.

leehyunk6310 committed:
feat : DB 연결 설정

leehyunk6310 committed:
feat: 구글 로그인 구현

leehyunk6310 committed:
feat: 카카오, 네이버 로그인 구현

leehyunk6310 committed:
chore: add _ to parameter accessToken, refreshToken

leehyunk6310 committed:
feat: 로그인 시 CreateUserDto 반환

leehyunk6310 committed:
Merge branch 'main' into 25-social-login

leehyunk6310 committed:
feat: 소셜 회원가입

- user.dto.ts -> auth.dto.ts 변경
- user.service.ts 합치기
- 임시 회원가입시 토큰 생성
- 토큰 시간 10분으로 변경
- cors 허용

leehyunk6310 committed:
feat: 로그인 시 jwt를 쿠키로 보내줌

leehyunk6310 committed:
Merge branch 'main' into 25-social-login

leehyunk6310 commented on src/auth/auth.controller.ts:
```diff
@@ -11,30 +11,162 @@ import {
+  ) {}
 
-  // 구글 로그인
+  @Public()
+  @Post('/sign-up')
+  async signUp(@Req() req: Request, @Body() authSignUpDto: AuthSignUpDto) {
+    const [_type, token] = req?.headers?.authorization?.split(' ') ?? [];
+
+    if (isNil(token)) {
+      throw new UnauthorizedException('need token');
+    }
+
+    try {
+      const payload: AuthLoginDto = await this.jwtService.verifyAsync(token, {
+        secret: this.configService.get<string>('JWT_SECRET'),
+      });
+
+      if (
+        !isNil(await this.userService.findUserByProviderId(payload.providerId))
+      ) {
+        throw new Error('User Already Registered');
+      }
+
+      const user = await this.authService.makeUser(payload, authSignUpDto);
+
+      this.authService.signUp(user);
+    } catch (e) {
+      if (e instanceof TokenExpiredError) {
+        throw new UnauthorizedException('jwt expired');
+      }
+      console.error(e);
+    }
+  }
+
```
헤더의 토큰으로부터 얻은 payload에 providerId값을 가지고 DB를 확인한 후 회원이 없으면 회원가입을 진행시킵니다.

leehyunk6310 commented on src/auth/auth.controller.ts:
```diff
@@ -11,30 +11,162 @@ import {
+      });
+      res.redirect('/to-front/fail');
+      return;
+    }
```
모든 콜백에 대하여 같은 방식으로 작성하였습니다.
provider, providerId, email값을 임시 토큰에 저장하여 추후 회원가입에 사용하였습니다.
회원이 존재하는 경우 signIn2 함수를 사용하여 accessToken을 발급하여 프론트에 넘겨줍니다.

leehyunk6310 commented on src/auth/auth.service.ts:
```diff
@@ -26,7 +28,54 @@ export class AuthService {
+  async signUp(user: User) {
+    this.userService.signUp(user);
+  }
+
```
임시 회원가입 함수로 provider, providerId, email을 임시 토큰에 저장합니다.

leehyunk6310 commented on src/auth/auth.controller.ts:
```diff
@@ -11,30 +11,162 @@ import {
   @UseGuards(AuthGuard('google'))
-  async callbackGoogle(@Req() req: Request, @Res() _res: Response) {
-    const user = req.user;
+  async callbackGoogle(@Req() req: Request, @Res() res: Response) {
+    if (req.user) {
+      const { provider, providerId, email } = req.user as AuthLoginDto;
+
+      const user = await this.userService.findUserByProviderId(providerId);
+
+      if (user) {
+        const accessToken = this.authService.signIn2(providerId);
+        res.cookie('Authentication', (await accessToken).accessToken, {
+          httpOnly: true,
+          secure: false,
+        });
+        res.redirect('/to-front/success');
+        return;
+      }
+
+      const tmp_token = this.authService.tempSignUp(
+        provider,
+        providerId,
+        email,
+      );
+      res.cookie('info', (await tmp_token).tmpToken, {
+        httpOnly: true,
+        secure: false,
+      });
+      res.redirect('/to-front/fail');
+      return;
+    }
+  }
```
모든 콜백에 대하여 같은 방식으로 작성하였습니다.
provider, providerId, email값을 임시 토큰에 저장하여 추후 회원가입에 사용하였습니다.
회원이 존재하는 경우 signIn2 함수를 사용하여 accessToken을 발급하여 프론트에 넘겨줍니다.

leehyunk6310 commented on src/auth/auth.service.ts:
```diff
@@ -26,7 +28,54 @@ export class AuthService {
     };
   }
+
+  async tempSignUp(provider: string, providerId: string, email: string) {
+    const payload = {
+      provider: provider,
+      providerId: providerId,
+      email: email,
+    };
+
+    return {
+      tmpToken: await this.jwtService.signAsync(payload),
+    };
+  }
```
임시 회원가입 함수로 provider, providerId, email을 임시 토큰에 저장합니다.

leehyunk6310 commented on src/auth/auth.service.ts:
```diff
@@ -26,7 +28,54 @@ export class AuthService {
+  }
+
+  // 이름 변경 필요
+  async signIn2(providerId: string) {
+    const user = await this.userService.findUserByProviderId(providerId);
+
+    if (isNil(user)) {
+      throw new UnauthorizedException();
+    }
+
+    const payload = {
+      id: user.id,
+      roles: user.role,
+    };
+    return {
+      accessToken: await this.jwtService.signAsync(payload),
+    };
+  }
```
로그인 함수로, 기존 signIn과 겹치기에 이름을 변경할 필요가 있습니다.

vimkim replied:
기존 signIn 은 제가 mockSignIn으로 바꿔두겠습니다

leehyunk6310 replied:
이 부분은 그럼 맡기겠습니다.

leehyunk6310 commented on src/user/user.service.ts:
```diff
@@ -25,7 +31,26 @@ export class UserService {
   ];
 
-  async findOne(username: string): Promise<User | undefined> {
+  async findOne(username: string): Promise<Users | undefined> {
     return this.users.find((user) => user.username === username);
   }
```
User 엔티티와 클래스 명이 겹쳐서 온메모리에서 사용하던 User를 Users로 변경하였습니다.

vimkim replied:
MockUser로 제가 바꿔두겠습니다. 말씀하시고 지우셔도 됩니다!

leehyunk6310 replied:
이 부분도 맡기겠습니다

leehyunk6310 committed:
rename: tmp_token -> tmpToken으로 변경

vimkim commented:
닫을 이슈 연결해주시면 좋을 거 같아요! 이번에는 제가 해드리겠습니다

leehyunk6310 commented:
깜빡했네요! 감사합니다 다음부터 풀리퀘할때 확인하겠습니다 !

vimkim commented on src/auth/auth.controller.ts:
```diff
@@ -11,30 +11,158 @@ import {
+          httpOnly: true,
+          secure: false,
+        });
+        res.redirect('/to-front/success');
+        return;
+      }
+
+      const tmpToken = this.authService.tempSignUp(provider, providerId, email);
+      res.cookie('info', (await tmpToken).tmpToken, {
+        httpOnly: true,
+        secure: false,
+      });
+      res.redirect('/to-front/fail');
+      return;
+    }
```
/to-front 대신

.env에
$CLIENT_URL 변수를 만드시고
localhost:5173 을 넣어두시면 좋을 거 같아요

leehyunk6310 replied:
수정했습니다

vimkim commented on src/auth/auth.controller.ts:
```diff
@@ -11,30 +11,158 @@ import {
+      const { provider, providerId, email } = req.user as AuthLoginDto;
+
+      const user = await this.userService.findUserByProviderId(providerId);
+
+      if (user) {
+        const accessToken = this.authService.signIn2(providerId);
+        res.cookie('Authentication', (await accessToken).accessToken, {
+          httpOnly: true,
+          secure: false,
+        });
+        res.redirect('/to-front/success');
+        return;
+      }
+
+      const tmpToken = this.authService.tempSignUp(provider, providerId, email);
+      res.cookie('info', (await tmpToken).tmpToken, {
+        httpOnly: true,
+        secure: false,
+      });
+      res.redirect('/to-front/fail');
+      return;
+    }
```
쿠키를 붙이는  공통 로직이 3번 반복되는 거 같은데 따로 함수로 빼면 유지 보수에 편할 거 같습니다!

leehyunk6310 replied:
수정했습니다 !

vimkim commented on src/auth/auth.service.ts:
```diff
@@ -26,7 +28,54 @@ export class AuthService {
+  async makeUser(authLoginDto: AuthLoginDto, authSignUpDto: AuthSignUpDto) {
+    const user = new User();
+
+    user.platform = authLoginDto.provider;
+    user.providerId = authLoginDto.providerId;
+    user.email = authLoginDto.email;
+    user.nickname = authSignUpDto.nickname;
+    user.gender = authSignUpDto.gender;
+    user.job = authSignUpDto.job;
+    user.birthdate = authSignUpDto.birthDate;
```
user 엔티티에도 Dto에 적힌대로 platform -> provider
birthdate -> birthDate 이렇게 통일하면 좋을 것 같습니다.

leehyunk6310 replied:
수정했습니다 !

vimkim commented on src/auth/auth.service.ts:
```diff
@@ -26,7 +28,54 @@ export class AuthService {
+      throw new UnauthorizedException();
+    }
+
+    const payload = {
+      id: user.id,
+      roles: user.role,
```
서비스에 있는 예시 유저 엔티티에 제가 role이라고 했었나요?
제가 네이밍을 잘못 했나보네요..
 user.role 보다는 user.roles가 나을 것 같네요...

leehyunk6310 replied:
수정했습니다 !

vimkim commented on src/auth/strategies/google.strategy.ts:
```diff
@@ -26,11 +25,7 @@ export class GoogleStrategy extends PassportStrategy(Strategy, 'google') {
-      name: name?.givenName,
-      email,
-    };
+    const user: AuthLoginDto = new AuthLoginDto('google', id, email!);
+    return user;
```
email 뒤에 느낌표 없앨 방법이 없을까요?

leehyunk6310 replied:
수정했습니다

vimkim commented:
고생하셨습니다. 댓글 내용 확인하시고 답변 주시면 어셉하겠습니다!

leehyunk6310 committed:
fix: 요청사항 수정

- controller 코드 내용  service로 이전
- strategy email에 Nil값 처리
- entity 필드명 수정(provider, roles, birthDate)

leehyunk6310 committed:
fix: redirect url 수정

- success, new-user앞에 login추가
- 컨트롤러에서 사용하지 않는 서비스 생성자에서 제외
